### PR TITLE
`RegexFSM`: Cache Legal-Token Mask as `torch.Tensor` to Improve Performance

### PR DESCRIPTION
**Pull request by lapp0**
Fixes https://github.com/outlines-dev/outlines/issues/1011

Fixes https://github.com/outlines-dev/outlines/issues/920

## Problem

Runtime performance for structured generation was weak because the legal next token mask is `List[int]` in main. This is inefficient because we have to convert a list of potentially tens of thousands of integers to `torch.tensor`.

## Solution

Cache legal next token ids as a `torch.tensor`. Don't construct the mask from `List[int]`, construct it from `torch.Tensor`.

With a vocabulary of ~50,000 tokens, worst case performance decreases 50-fold.

```
| Change   | Before [833f68f8]    | After [d9baa1ba]    |   Ratio | Benchmark (Parameter)                                                                            |
|----------|----------------------|---------------------|---------|--------------------------------------------------------------------------------------------------|
| -        | 56.0±0.3ms           | 1.05±0.02ms         |    0.02 | bench_processors.LogitsProcessorStructuredBenchmark.time_structured_generation('numpy', '[^Z]*') |
| -        | 58.1±0.6ms           | 1.05±0.01ms         |    0.02 | bench_processors.LogitsProcessorStructuredBenchmark.time_structured_generation('torch', '[^Z]*') |
```

### CODE REVIEW

**Comment by Qubitium** [comment]
@lapp0 Do you expect even more gains for models with >200k tokenizers as it has become more common?

**Comment by lapp0** [comment]
@Qubitium yes, the gains are proportional to the size of the tokenizers vocabulary.

**Comment by rlouf** [comment]
Great catch!
